### ISSUE TITLE
extend support for custom SQL functions (mutations, other functions types)

### ISSUE THREAD

**Issue by dsandip** [hasura/graphql-engine#1514]
1. Support for SQL functions in mutations
2. Support other function types

**Comment by revskill10** [comment]
@dsandip Is it possible to support `pg notify/listen` inside custom mutation, so that we could have custom subscription via `pg notify/listen` ?

**Comment by sastraxi** [comment]
Would this include exposing `volatile` functions as mutations? The work-around I've found so far is to have a work log per function I want to call and simply use the event system's `INSERT` trigger to webhook out to a process which will carry out the desired mutation.

Apologies if I misunderstood this issue (in which case I'll open a new one).

**Comment by MhdSyrwan** [comment]
@shahidhk any updates on this (since I saw that you were planning to implement it)?

> @revskill10 We'll add them in the coming iterations.

_Originally posted by @shahidhk in https://github.com/hasura/graphql-engine/pull/1073#issuecomment-441159431_

Thanks in advance.

**Comment by shahidhk** [comment]
@MhdSyrwan There is no exact ETA at the moment.

**Comment by tafelito** [comment]
Is there anyway to track the function that are manually created right now? Selecting the track option won't work as it says the function does not return a "COMPOSITE" type and if I refresh the browser I lose all the declarations

**Comment by marionschleifer** [comment]
@tafelito you can't return a composite type from a function. A function has to return a SETOF table.

**Comment by tafelito** [comment]
@marionschleifer so if I have something like this

```
-- Create the function
CREATE OR REPLACE FUNCTION insert_license()
  RETURNS trigger AS $BODY$
DECLARE licenses_avail INTEGER;
BEGIN
 SELECT a.licenses_avail INTO licenses_avail FROM "clients_codes" a WHERE a.id = NEW."client_code";
 IF licenses_avail <= 0 THEN
     RAISE EXCEPTION 'No more licenses available';
 END IF;
 RETURN NEW;
END;
$BODY$ LANGUAGE plpgsql;

--- Create the trigger with the above function
CREATE TRIGGER insert_license BEFORE INSERT ON "licenses" FOR EACH ROW EXECUTE PROCEDURE insert_license();
```

I don't have a SETOF table to return. Is there any other way I can keep track of this functions?

**Comment by shahidhk** [comment]
@tafelito This is tracked in the Postgres schema as a trigger and the associated function. You don't want Hasura to "track" this. 

Hasura's tracking is for functions which you want to expose as GraphQL queries.

**Comment by tafelito** [comment]
So if I need to see the functionsI created or if I want to edit one I have to connect directly to Postgress outside Hasura? Is that what you're saying?

**Comment by shahidhk** [comment]
Yes, since it has nothing to do with Hasura or GraphQL. You can use the Run SQL window in Hasura also. Just don't tick the "Track this" box.

**Comment by mnesarco** [comment]
Hi Friends, I am not sure if this is offtopic, but I managed to call PLPgsql Functions as mutations with Hasura 1.0 and I want to share my approach with you:

https://mnesarco.github.io/blog/2020/02/25/hasura-postgresql-function-as-mutation

**Comment by MhdSyrwan** [comment]
@mnesarco your approach is great and self-integrated. it allows developers even to inspect and trace previous calls and their responses! :+1: 

Question: I didn't get this part `INSERT INTO action('sum');` how does it work without writing "VALUES" before `('sum')`?
http://sqlfiddle.com/#!17/11e102/3

BTW, for people who want a simple quick solution, they may just create a view with a trigger to replace the create action of it in order to make a single callable function.

**Comment by mnesarco** [comment]
Hi @MhdSyrwan thanks for your review. I have fixed the missing "values" part.
I do not understand the view with trigger idea, can you explain it a little?

**Comment by MhdSyrwan** [comment]
Example:

You can create a dummy view like
```sql
CREATE VIEW my_action AS SELECT 'dummy';
```

Then write a trigger to implement its INSERT action like this:
```sql
CREATE OR REPLACE FUNCTION my_action_replace_insert ()
  RETURNS TRIGGER
  LANGUAGE plpgsql
  AS $function$
DECLARE
BEGIN
  -- function body here
END;
$function$;

CREATE TRIGGER my_action_replace_insert_trigger
    INSTEAD OF INSERT ON "my_action"
    FOR EACH ROW
    EXECUTE PROCEDURE public.my_action_replace_insert ();
```

Then, by tracking the `my_action` view in Hasura UI, you can call the function by calling mutation `insert_my_action`.

**Comment by mnesarco** [comment]
I think that the view approach does not allow to return any result from the function but I can be wrong. I will try your idea and if it works i will add a section at the end of the article.

**Comment by MhdSyrwan** [comment]
You're right. it does not allow. Actually I didn't care about this feature since the whole point of this issue was to make a custom change in the database rather than to return something.

Functions that return value without altering DB are already supported check https://github.com/hasura/graphql-engine/pull/1073.

Still, your approach would be the best for somebody trying to make a change and get some info at the same time.

**Comment by mnesarco** [comment]
Yes my solution is for calling functions without write restrictions, without input restrictions and without returning restrictions.

**Comment by mnesarco** [comment]
I have published a reference source code here: https://github.com/mnesarco/hasura-pg-actions
It could be useful for someone.

**Comment by RodolfoSilva** [comment]
@mnesarco. I think when actions is arrived we will have this feature built in. Right now we can do this with external functions. But you approach is very useful, but we need some permission checks before expose this for the users. I believe this is the way.

**Comment by mnesarco** [comment]
Hi @RodolfoSilva Thanks for your review. I am using this approach in a real Application now and I am very happy, it works completely inside postgres without calling any external endpoint and it is fully transactional. It would be great to have this functionality built in. I am aware of the current hasura 1.2-beta actions, but it requires external endpoints.

**Comment by pkvince** [comment]
Thank you for sharing your solution @mnesarco ! I have been waiting for a solution like this to come up for a while... Trying this right away!

**Comment by AtjonTV** [comment]
Is there any ETA on this?

I want to use Hasura as a Backend for my app, but I have the issue that I want to calculate the SUM of multiple columns on the fly using SQL Functions so that the client just needs to display the value.

The issue I am having is that the function is `volatile`. And it needs to be, as the rows could change, even though you cant do it in the App right now. But it has to be supported (for error correction).

Basically I have a table of Bills where each Tax Group has its own Column (In Austria we have 5 groups, so 5 columns) and I want to calculate the total sum of a bill. This boils down to sum up the 5 money typed fields, and returning the result.

The client doesnt need to know the split-up values yet.

**Comment by karibertils** [comment]
I have a case where using a VOLATILE SQL function in a computed field would simplify things alot.

I also noticed I can create the function as STABLE, add it as computed field, and then change the function to VOLATILE and it works just fine in hasura.

But feels hacky and will probably break on something like metadata reload. Also seem's there's not really much needed to support this except allowing VOLATILE function to be tracked, at least in my case.

Will the work for this issue support this use case as well ?

**Comment by Bessonov** [comment]
@karibertils look at linked draft pull request.

**Comment by karibertils** [comment]
@Bessonov I can see in the pull request that VOLATILE functions will be supported as mutations.

But I can't really tell if adding those mutations as computed fields will also be supported like with the STABLE functions ?

It might be a bit odd behaviour doing query { table { computed } } that's actually mutating data before returning the results.

**Comment by jberryman** [comment]
@karibertils There is no plan at the moment to allow `VOLATILE` functions to exist under the `Query` root field; the thinking is that because the graphql spec requires fields (except top-level mutation fields) to be idempotent and side-effect free, allowing a VOLATILE function would result in a graphql API that violates the spec (and e.g. causes issues with caching etc)

Can you give me an idea of your use-case? Maybe you have a function that is "morally" side-effect -free, but needs to be marked VOLATILE?

Also fwiw we also have some early plans to simplify/unify "function tracking" and "computed fields" APIs which should at least make this easier to talk about

**Comment by ayuryshev** [comment]
@jberryman Ability to query VOLATILE functions would open possiblity to implement server-side memoization.
Where only side-effect is cached calculation. 
IDK is it breaks graphql specification but that would be great in practice.

It was several times when I started to think "aha ... and we can cache result so it .. Stop - hasura will no allow it.. Sad"

**Comment by jberryman** [comment]
@ayuryshev thanks! Yep the caching use-case you describe makes sense to me. I don't think that will happen in the VOLATILE function-as-mutations work, since it requires some discussion. Would you mind opening a feature request issue and mention your use case? It would also be helpful if you wanted to link to resources that describe the technique you're using.

**Comment by jberryman** [comment]
PR is in review ^

**Comment by revskill10** [comment]
Is this released in 1.4 ?

**Comment by hthillman** [comment]
Any update on when this will be released?

**Comment by ayuryshev** [comment]
> @ayuryshev thanks! Yep the caching use-case you describe makes sense to me. I don't think that will happen in the VOLATILE function-as-mutations work, since it requires some discussion. Would you mind opening a feature request issue and mention the use-case you just outlined? It would also be helpful if you wanted to link to resources that describe the technique you're using.

@jberryman 
Please take a look on: https://github.com/hasura/graphql-engine/issues/6333#issue-763379472

**Comment by karibertils** [comment]
@jberryman 

My use case for VOLATILE SQL function in query is as follows.

For every user there should always be exactly one in-progress order. If I query ```user(id: X) { current_order { ...order_details } }``` it should return that order.

If the user has no orders, or has only completed orders, it would create new empty order and return that as the current_order. If the user has existing in-progress order it would return the existing one. The intention is to simplify the API usage. Skipping the need to query, mutate, query again in cases with no current_order.

The orders can get finished in a few different ways, and feels less error-prone handling it at a single place rather than tracking it in many. I know this can be accomplished using SQL triggers or in the backend etc. But having a VOLATILE SQL function + computed field handle this felt like clean solid way.

**Comment by jberryman** [comment]
`VOLATILE` functions are supported as queries on master (see docs and linked PR). Apologies for not updating the discussion here, and thanks for the input!

**Comment by lancedouglas1** [comment]
Well done team Hasura.

Volatile functions are more valuable to the users of your product than the dogma of graphql abstaining from them.

I see this as an inflection point in customer value over technocracy that will serve Hasura well into being a global enterprise suite.

**Comment by shashankrepo** [comment]
> You're right. it does not allow. Actually I didn't care about this feature since the whole point of this issue was to make a custom change in the database rather than to return something.
> 
> Functions that return value without altering DB are already supported check #1073.
> 
> Still, your approach would be the best for somebody trying to make a change and get some info at the same time.

You can still return the value you want using your method @MhdSyrwan @mnesarco . https://hasura.io/blog/postgres-triggers-on-graphql-mutations-682bf48db023/

`CREATE OR REPLACE FUNCTION my_action_replace_insert ()
  RETURNS TRIGGER
  LANGUAGE plpgsql
  AS $function$
DECLARE
BEGIN
  -- function body here
**RETURN NEW**
END;
$function$;

CREATE TRIGGER my_action_replace_insert_trigger
    INSTEAD OF INSERT ON "my_action"
    FOR EACH ROW
    EXECUTE PROCEDURE my_action_replace_insert ();`
